perllib/FixMyStreet/Email.pm

```diff
@@ -121,41 +118,152 @@ sub send_cron {
-    # $attachments should be an array_ref of things that can be parsed to Email::MIME,
-    # for example
-    #    [
-    #      body => $binary_data,
-    #      attributes => {
-    #          content_type => 'image/jpeg',
-    #          encoding => 'base64',
-    #          filename => '1234.1.jpeg',
-    #          name     => '1234.1.jpeg',
-    #      },
-    #      ...
-    #    ]
-    #
-    # XXX: mySociety::Email::construct_email isn't using a MIME library and
-    # requires more analysis to refactor, so for now, we'll simply parse the
-    # generated MIME and add attachments.
-    #
-    # (Yes, this means that the email is constructed by Email::Simple, munged
-    # manually by custom code, turned back into Email::Simple, and then munged
-    # with Email::MIME.  What's your point?)
-
-    require Email::MIME;
-    my $mime = Email::MIME->new($message);
-    $mime->parts_add([ map { Email::MIME->create(%$_)} @$attachments ]);
-    my $data = $mime->as_string;
-
-    # unsure why Email::MIME adds \r\n. Possibly mail client should handle
-    # gracefully, BUT perhaps as the segment constructed by
-    # mySociety::Email::construct_email strips to \n, they seem not to.
-    # So we re-run the same regexp here to the added part.
-    $data =~ s/\r\n/\n/gs;
-
-    return $data;
+=item construct_email SPEC
+
+Construct an email message according to SPEC, which is an associative array
+containing elements as given below. Returns an Email::MIME email.
+
+=over 4
+
+=item _template_, _parameters_
+
+Templated body text and an associative array of template parameters. _template
+contains optional substititutions <?=$values['name']?>, each of which is
+replaced by the value of the corresponding named value in _parameters_. It is
+an error to use a substitution when the corresponding parameter is not present
+or undefined. The first line of the template will be interpreted as contents of
+the Subject: header of the mail if it begins with the literal string 'Subject:
+' followed by a blank line. The templated text will be word-wrapped to produce
+lines of appropriate length.
+
+=item _attachments_
+
+An arrayref of hashrefs that can be passed to Email::MIME.
+
+=item To
+
+Contents of the To: header, as a literal UTF-8 string or an array of addresses
+or [address, name] pairs.
+
+=item From
+
+Contents of the From: header, as an email address or an [address, name] pair.
+
+=item Cc
+
+Contents of the Cc: header, as for To.
+
+=item Reply-To
+
+Contents of the Reply-To: header, as for To.
+
+=item Subject
+
+Contents of the Subject: header, as a UTF-8 string.
+
+=item I<any other element>
+
+interpreted as the literal value of a header with the same name.
+
+=back
+
+If no Date is given, the current date is used. If no To is given, then the
+string "Undisclosed-Recipients: ;" is used. It is an error to fail to give a
+templated body, From or Subject (perhaps from the template).
+
+=cut
+sub construct_email ($) {
```

Comment:
I guess you included this here because other projects rely on the version in commonlib? Would it make sense to keep this within commonlib under a new name, then other dependent code can be refactored over time and eventually the old `construct_email` code removed? It'd cut down on duplication over the long run.